design-proposals/klip-66-multi-param-udafs.md

```diff
@@ -0,0 +1,165 @@
+# KLIP 66 - Multiple parameter UDAFs
+
+**Author**: @jzaralim | 
+**Release Target**: TBD | 
+**Status**: In Discussion | 
+**Discussion**: _link to the design discussion PR_
+
+**tl;dr:** _Enable users to create UDAFs with multiple parameters._
+
+## Motivation and background
+
+Currently, UDAFs can only accept a single column to aggregate. Users have shown interest
+([#5747](https://github.com/confluentinc/ksql/issues/5747), [#7435](https://github.com/confluentinc/ksql/issues/7435), [#5300](https://github.com/confluentinc/ksql/issues/5300))
+in aggregating over multiple columns. The workaround is to combine all of the required columns into a `Struct` and then implementing a `UDAF` to
+work with `Struct` with that specific schema. This is tedious, and requires the user to implement a new version of their function for every schema they need to work with.
+
+## What is in scope
+
+* UDAFs that accept variadic parameters: `example(col_int_1, col_int_2, col_int_3...) -> result`
+* UDAFs that accept a specific number of parameters, potentially of different types: `example(col_int, col_string) -> result`
+* Variadic TopK - a version of `TOPK` that returns the top K values of one column, along with the values of other columns from those records.
```

Comment:
What is the intended return type for TopK?  Would it be a struct containing the other columns?

Reply:
Yeah, a struct. We're not outputting multiple columns in this KLIP.